plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs

```diff
@@ -0,0 +1,138 @@
+  , redeem
+  , redeemTx
+  , balance
+  , address
+  , newAccount
+  ) where
+
+import           Control.Lens
+import qualified Data.Map                                          as Map
+import           Data.Maybe                                        (fromMaybe)
+
+import           Language.Plutus.Contract
+import qualified Language.PlutusTx                                 as PlutusTx
+
+import qualified Language.Plutus.Contract.Typed.Tx                 as TypedTx
+import           Ledger                                            (Address, PubKey, TxOut)
+import qualified Ledger                                            as Ledger
+import           Ledger.TxId                                       (TxId)
+import           Ledger.Typed.Scripts                              (ScriptType (..))
+import qualified Ledger.Typed.Scripts                              as Scripts
+import qualified Ledger.Validation                                 as V
+import           Ledger.Value                                      (CurrencySymbol, TokenName, Value)
+import qualified Ledger.Value                                      as Value
+
+import qualified Language.PlutusTx.Coordination.Contracts.Currency as Currency
+
+newtype AccountOwner = AccountOwner { unAccountOwner :: (CurrencySymbol, TokenName) }
+    deriving newtype Eq
+
+data TokenAccount
+
+instance ScriptType TokenAccount where
+    type RedeemerType TokenAccount = ()
+    type DataType TokenAccount = AccountOwner
+
+{-# INLINEABLE accountToken #-}
+accountToken :: AccountOwner -> Value
+accountToken (AccountOwner (symbol, name)) = Value.singleton symbol name 1
+
+{-# INLINEABLE validate #-}
+validate :: AccountOwner -> () -> V.PendingTx -> Bool
+validate owner _ ptx = V.valueSpent ptx `Value.geq` accountToken owner
+
+scriptInstance :: Scripts.ScriptInstance TokenAccount
+scriptInstance =
+    Scripts.Validator @TokenAccount $$(PlutusTx.compile [|| validate ||]) $$(PlutusTx.compile [|| wrap ||])
+    where
+    wrap = Scripts.wrapValidator @AccountOwner @()
+
+address :: Address
+address = Scripts.scriptAddress scriptInstance
+
+-- | Pay some money to the given token
+pay :: (AsContractError e, HasWriteTx s) => AccountOwner -> Value -> Contract s e TxId
+pay owner vl =
+    let ds = Ledger.DataScript (PlutusTx.toData owner)
+    in writeTxSuccess (payToScript vl address ds)
+
+ownsTxOut :: AccountOwner -> TxOut -> Bool
+ownsTxOut owner txout =
+    Just owner == (Ledger.txOutData txout >>= PlutusTx.fromData . Ledger.getDataScript)
+
+-- | Create a transaction that spends all outputs belonging to the 'AccountOwner'.
+redeemTx
+    :: ( HasUtxoAt s )
+    => AccountOwner
+    -> PubKey
+    -> Contract s e UnbalancedTx
+redeemTx owner pk = do
+    utxos <- utxoAt (Scripts.scriptAddress scriptInstance)
+    let filterByOwner _ = ownsTxOut owner
+        tx = TypedTx.collectFromScriptFilter filterByOwner utxos scriptInstance ()
+    -- TODO. Replace 'PubKey' with a more general 'Address' type of output?
```

Comment:
Not sure I understand what you want generalize? Is it that you want to be able to send it to a script?

Reply:
At the moment the two constraints "token X must be present" and "amount X must be paid to Y" are indistinguishable in `UnbalancedTx` and the comment was meant to suggest that there should be a way of saying that a token (value) must be moved by the transaction, without saying anything about the outputs it comes from or goes to.

Right now "token X must be present" is implemented as a pubkey output with a value of exactly X, but this means that (a) we need to know which pubkey to use for the output in `redeemTx`, which isn't a concern of the contract itself, and (b) yes we can't send it to a script. Although if we had the "token X must be present" constraint then (b) wouldn't matter anymore.

I would really like to be able to write these contracts without any reference to public key or script addresses (as far as ownership is concerned), just tokens.